library-best-practices/JLBP-6.md

```diff
@@ -0,0 +1,165 @@
+major version 2, a choice needs to be made: to rename or not rename? This
+question applies to both items listed above, the Maven ID (1) and the Java
+package (2).
+
+Recommendations:
+
+- *If* the new library surface is delivered under a new Java package, either use
+  a different Maven ID (different group ID or artifact ID) or bundle the old and
+  new packages together in the under the original Maven ID.
+- *If* the breaking change is made in-place and the Java package is kept the
+  same, use the same Maven ID (same group ID and same artifact ID).
+- There is a tradeoff: if an extremely large library (for example thousands of
+  classes) breaks a single method, the cost of a package rename will be higher
+  than dealing with the fallout of dealing with diamond dependency conflicts for
+  that particular method. The larger the portion of a library that breaks
+  between major versions, the higher the value is from renaming the Java
+  package.
+- Whether making breaking changes or making no breaking changes, don't publish
+  the same classes under multiple Maven IDs. Or put another way: Once you have
+  published under a particular Maven ID, you are stuck with it until you rename
+  your Java package.
+
+Extended discussion
+-------------------
+
+Consider the following renaming scenario:
+
+- The Maven coordinates start as `g1:a1:1.0.0`
+  - With no rename of the Maven ID, the new major version's Maven coordinates
+    are `g1:a1:2.0.0`. (Only one of the versions will be present on the
+    classpath even if both are in the dependency tree.)
+  - With a rename of the Maven ID, the new major version's Maven coordinates are
+    `g2:a2:2.0.0`. (Both versions can be present on the classpath.)
+- The Java package starts as `com.google.a1`
+  - With no rename, the new major version's Java package is still
+    `com.google.a1`.
+  - With a rename, the new major version's Java package is `com.google.a2`.
+
+Given this scenario, here are the possible combinations of renamings:
+
+- **Don't rename Java package**:
+  - **Don't rename Maven ID (case 1)**: This approach can result in diamond
+    dependency conflicts because different branches of a dependency tree can
+    depend on different major versions, and the build system (Maven or Gradle)
+    will only choose one of them to use. Users are forced to change their code
+    between versions, but only for library surface that was changed.
+  - **Rename Maven ID (case 2)**: Users can easily pull in both the old jar
+    (`g1:a1:1.0.0`) and the new jar (`g2:a2:2.0.0`) accidentally through
+    transitive dependencies, because Maven treats the two artifacts as distinct,
+    but since they have classes with the same Java packages, they are loaded in
+    unpredictable ways, leading to runtime exceptions (for example
+    `ClassNotFoundException`). It is difficult for users to ensure that their
+    build tree only includes one of the artifacts. **NEVER DO THIS**.
```

Comment:
I think the servlet API did exactly this. Was there a good reason we're missing here (e.g. provided jars) or did they screw up?

Reply:
Based on my (limited) understanding, they screwed up - now we have to go around advising everyone (e.g. in JLBP-5) to use `provided` as the scope for servlet-api, otherwise people will see conflicts.

Reply:
Servlet API _needs_ to be provided by the servlet container itself. That's the only sane way for the container to communicate with the servlets themselves. It should always be `provided`, in the same way any AppEngine APIs and Android APIs would be `provided`; they are part of the runtime. But all of those are different from normal libraries.